arch/x86/pagetables.c

```diff
@@ -92,6 +92,11 @@ static inline void dump_pte(void *entry, mfn_t table, int level, int index) {
            level, index, paddr, flags);
 }
 
+static inline int canonical_va(void *va) {
+    unsigned long sign = (_ul(va) >> (VA_BITS - 1)) & 1;
+    return sign * (_ul(1) << (sizeof(va) * 8 - VA_BITS)) - 1 == _ul(va) >> VA_BITS;
```

Comment:
This looks a little convoluted and won't do the right thing for user addresses, i.e. VAs with the extended sign bit set to zero.

What about something more simple like this?:
```c
static inline bool is_canon_va(const void *va) {
    const unsigned int sign_bits = BITS_PER_LONG - VA_BITS;
    return _ptr(((long)va << sign_bits) >> sign_bits) == va;
}
```

Reply:
I agree it looks pretty hard to read.. Will use `BITS_PER_LONG`.
I think your suggestion won't catch high addresses?
Remember that KTF allows usermode to allocate high addresses.

For user addresses, what we want is `0 * 0xffff == adr >> 48`
For kernel/high addresses, we want `1 * 0xffff == adr >> 48`

Reply:
I guess something like 
```c
static inline bool is_canon_va(void *va) {
    unsigned sign = (_ul(va) >> (VA_BITS - 1)) & 1;
    return sign * (1 << (BITS_PER_LONG - VA_BITS)) - 1 == _ul(va) >> VA_BITS;
}
```

is getting a bit better?

Reply:
oh! i didn't realize the sign extension trick with the long cast! will fix!